mesecons_luacontroller/init.lua

```
@@ -259,49 +260,69 @@ local function remove_functions(x)
 	return x
 end
 
-local function get_interrupt(pos)
+-- itbl: Flat table of functions to run after sandbox cleanup, used to prevent various security hazards
+local function get_interrupt(pos, itbl)
 	-- iid = interrupt id
 	local function interrupt(time, iid)
+		-- NOTE: This runs within string metatable sandbox, so don't *rely* on anything of the form (""):y
+		-- Hence the values get moved out. Should take less time than original, so totally compatible
 		if type(time) ~= "number" then return end
-		local luac_id = minetest.get_meta(pos):get_int("luac_id")
-		mesecon.queue:add_action(pos, "lc_interrupt", {luac_id, iid}, time, iid, 1)
+		table.insert(itbl, function ()
+			-- Outside string metatable sandbox, can safely run this now
+			local luac_id = minetest.get_meta(pos):get_int("luac_id")
+			-- Check if IID is dodgy, so you can't use interrupts to store an infinite amount of data.
+			-- Note that this is safe from alter-after-free because this code gets run after the sandbox has ended.
```

Yes it can’t be made arbitrarily big, but still IID can be changed after call to `interrupt`.
BTW, didn’t know that tables could be used as IID.

---

Neither did I. Also note that functions being put in `itbl` only run after the sandbox ends - I.E. after the code is unable to change it.

---

>  I.E. after the code is unable to change it.

Yes I understand that. That is, there is no security problem. The problem I noticed is just inconvenience: it’s highly unobvious that the table is captured by reference, esp. remembering the fact `digiline_send` does the opposite.

---

Also now it may fail silently (if IID is too large), that’s not very good.

---

That's more an unfortunate side effect of attempting to keep compatibility with `digiline_send`'s "return false if message too large" behavior, but also trying to avoid breaking compatibility by making the actual process of setting an interrupt use serialize.
Keep in mind last time someone tried these kinds of changes, they got reverted because of what amounted to event-counter compatibility (insane reason IK).

---

It may be better to complain loudly, e.g. showing an error message in the formspec. That may actually be a better (for players) way to report failures because it doesn’t require checking return value (the thing no one wants to do).